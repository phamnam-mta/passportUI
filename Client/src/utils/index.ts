export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
