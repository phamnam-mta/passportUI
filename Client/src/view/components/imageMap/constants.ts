export const LAYER_NAME = "name";
export const IMAGE_LAYER_NAME = "imageLayer";
export const TEXT_VECTOR_LAYER_NAME = "textVectorLayer";
export const POD_VECTOR_LAYER_NAME = "podVectorLayer";
export const TABLE_BORDER_VECTOR_LAYER_NAME = "tableBorderVectorLayer";
export const TABLE_ICON_VECTOR_LAYER_NAME = "tableIconVectorLayer";
export const TABLE_ICON_BORDER_VECTOR_LAYER_NAME = "tableIconBorderVectorLayer";
export const CHECKBOX_VECTOR_LAYER_NAME = "checkboxBorderVectorLayer";
export const LABEL_VECTOR_LAYER_NAME = "labelledVectorLayer";
export const DRAWN_REGION_LABEL_VECTOR_LAYER_NAME = "drawnRegionLabelledVectorLayer";
export const DRAWN_REGION_VECTOR_LAYER_NAME = "drawnRegionVectorLayer";

export const SELECTED_PROPERTY = "selected";
export const SELECTION_MARK_STATE = "selectionMarkState";
export const HIGHLIGHTED_PROPERTY = "highlighted";
export const FIELD_PROPERTY = "field";
export const COLOR_PROPERTY = "color";
export const BACKGROUND_COLOR_PROPERTY = "backgroundColor";
export const DASHED_PROPERTY = "dashed";
export const DISABLEHIGHLIGHT_PROPERTY = "disableHighlight";
