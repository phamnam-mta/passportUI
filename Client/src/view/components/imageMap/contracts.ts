/**
 * @name - Feature Category
 * @description - Defines types of feature
 * @member Checkbox - Checkbox
 * @member DrawnRegion - User drawn region
 * @member Label - User label
 * @member Text - OCR text
 */
export enum FeatureCategory {
    Checkbox = "checkbox",
    DrawnRegion = "region",
    Label = "label",
    Text = "text",
}

/**
 * @name - Region Type
 * @description - Defines the region type within the asset metadata
 * @member Point - Specifies a vertex
 * @member Polygon - Specifies a region as a multi-point polygon
 * @member Polyline - Specifies a region as a multi-point line
 * @member Square - Specifies a region as a square
 * @member Rectangle - Specifies a region as a rectangle
 */
export enum RegionType {
    Point = "POINT",
    Polygon = "POLYGON",
    Polyline = "POLYLINE",
    Rectangle = "RECTANGLE",
    Square = "SQUARE",
}

/**
 * @name - Region
 * @description - Defines a region within an asset
 * @member id - Unique identifier for this region
 * @member type - Defines the type of region
 * @member category - Defines the feature category
 * @member tags - Defines a list of tags applied to a region
 * @member points - Defines a list of points that define a region
 * @member boundingBox - Defines a list of points that forms a bounding box
 * @member value - Defines a normalized value for this region
 * @member pageNumber - Defines a page number for this region
 * @member isTableRegion - Defines a flag determine if the region belongs to a table
 * @member changed - Defines a flag if this region had been changed
 */
export interface IRegion {
    id: string;
    type: RegionType;
    category: FeatureCategory;
    tags: string[];
    points?: IPoint[];
    boundingBox?: IBoundingBox;
    value?: string;
    pageNumber: number;
    isTableRegion?: boolean;
    changed?: boolean;
}

/**
 * @name - Bounding Box
 * @description - Defines the tag usage within a bounding box region
 * @member left - Defines the left x boundary for the start of the bounding box
 * @member top - Defines the top y boundary for the start of the boudning box
 * @member width - Defines the width of the bounding box
 * @member height - Defines the height of the bounding box
 */
export interface IBoundingBox {
    left: number;
    top: number;
    width: number;
    height: number;
}

/**
 * @name - Point
 * @description - Defines a point / coordinate within a region
 * @member x - The x value relative to the asset
 * @member y - The y value relative to the asset
 */
export interface IPoint {
    x: number;
    y: number;
}
