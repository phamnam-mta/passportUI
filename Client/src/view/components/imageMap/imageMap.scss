@import "~@fluentui/react/dist/sass/References";

.map-wrapper {
    display: flex;
    flex-grow: 1;
    background-color: $ms-color-gray20;

    #map {
        width: 100%;
        height: 100%;
    }
}
