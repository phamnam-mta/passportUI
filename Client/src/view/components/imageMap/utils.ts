import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
