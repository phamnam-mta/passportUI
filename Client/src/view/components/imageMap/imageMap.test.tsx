import { mount, ReactWrapper } from "enzyme";
import * as React from "react";
import { DragPan } from "ol/interaction";
import { ImageMap } from "./imageMap";
import { convertRegionToFeature } from "./utils";
import { Extent } from "ol/extent";

describe("<ImageMap />", () => {
    const noop = () => {};

    const makeMockRegion = () => {
        return {
            id: "0.0348,0.2955,0.1084,0.2960,0.1084,0.3094,0.0354,0.3090",
            value: "mock-region",
        } as any;
    };

    const makeBrowserEvent = () => {
        const mockEvent = { target: {} };
        return { originalEvent: mockEvent };
    };

    const mockExtent: Extent = [0, 0, 300, 400];
    let baseProps;
    let mockHandleFinishFeatureSelect;
    beforeEach(() => {
        mockHandleFinishFeatureSelect = jest.fn();
        baseProps = {
            imageUri: "test.jpg",
            imageWidth: 300,
            imageHeight: 400,
            onMapReady: noop,
            featureStyler: noop,
            tableBorderFeatureStyler: noop,
            tableIconFeatureStyler: noop,
            tableIconBorderFeatureStyler: noop,
            checkboxFeatureStyler: noop,
            labelFeatureStyler: noop,
            drawRegionStyler: noop,
            drawnRegionStyler: noop,
            modifyStyler: noop,
            onFinishFeatureSelect: mockHandleFinishFeatureSelect,
        };
    });

    describe("Rendering", () => {
        it("should match snapshot", () => {
            const wrapper = mount(<ImageMap {...baseProps} />);
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("ComponentDidMount", () => {
        it("should initialize editor map.", () => {
            const props = { ...baseProps, initEditorMap: true };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(instance.map).not.toBeNull();
            expect(instance.imageLayer).not.toBeNull();
            expect(instance.textLayer).not.toBeNull();
            expect(instance.tableBorderLayer).not.toBeNull();
            expect(instance.tableIconBorderLayer).not.toBeNull();
            expect(instance.tableIconLayer).not.toBeNull();
            expect(instance.checkboxLayer).not.toBeNull();
            expect(instance.drawnRegionLayer).not.toBeNull();
            expect(instance.drawnLabelLayer).not.toBeNull();
            expect(instance.labelLayer).not.toBeNull();
            expect(instance.podLayer).not.toBeNull();

            expect(instance.dragPan).not.toBeNull();
            expect(instance.draw).not.toBeNull();
            expect(instance.dragBox).not.toBeNull();
            expect(instance.modify).not.toBeNull();
            expect(instance.snap).not.toBeNull();
        });

        it("should initialize predict map.", () => {
            const props = { ...baseProps, initPredictMap: true };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(instance.map).not.toBeNull();
            expect(instance.imageLayer).not.toBeNull();
            expect(instance.textLayer).not.toBeNull();
            expect(instance.labelLayer).not.toBeNull();
            expect(instance.podLayer).not.toBeNull();

            expect(instance.dragPan).not.toBeNull();
            expect(instance.draw).not.toBeNull();
            expect(instance.dragBox).not.toBeNull();
            expect(instance.modify).not.toBeNull();
            expect(instance.snap).not.toBeNull();
        });

        it("should initialize layout map.", () => {
            const props = { ...baseProps, initLayoutMap: true };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(instance.map).not.toBeNull();
            expect(instance.imageLayer).not.toBeNull();
            expect(instance.textLayer).not.toBeNull();
            expect(instance.tableBorderLayer).not.toBeNull();
            expect(instance.tableIconBorderLayer).not.toBeNull();
            expect(instance.tableIconLayer).not.toBeNull();
            expect(instance.checkboxLayer).not.toBeNull();
            expect(instance.drawnRegionLayer).not.toBeNull();
            expect(instance.drawnLabelLayer).not.toBeNull();
            expect(instance.labelLayer).not.toBeNull();
            expect(instance.podLayer).not.toBeNull();

            expect(instance.dragPan).not.toBeNull();
            expect(instance.draw).not.toBeNull();
            expect(instance.dragBox).not.toBeNull();
            expect(instance.modify).not.toBeNull();
            expect(instance.snap).not.toBeNull();
        });
    });

    describe("ComponentDidUpdate", () => {
        let wrapper: ReactWrapper;
        let instance;

        beforeEach(() => {
            const props = { ...baseProps, initEditorMap: true };
            wrapper = mount(<ImageMap {...props} />);
            instance = wrapper.instance() as any;
        });

        it("should update image when uri changed.", () => {
            const newImageUri = "another.jpg";
            const spySetImage = jest.spyOn(instance, "setImage");
            wrapper.setProps({ imageUri: newImageUri });
            expect(spySetImage).toBeCalledWith(newImageUri, instance.imageExtent);
        });

        it("should update image when angle changed.", () => {
            const spySetImage = jest.spyOn(instance, "setImage");
            wrapper.setProps({ imageAngle: 90 });
            expect(spySetImage).toBeCalledWith(baseProps.imageUri, instance.imageExtent);
        });

        it("should update interactions when setting draw region mode.", () => {
            const spyAddInteraction = jest.spyOn(instance, "addInteraction");
            const spyRemoveInteraction = jest.spyOn(instance, "removeInteraction");
            const spyInitializeDraw = jest.spyOn(instance, "initializeDraw");
            const spyInitializeModify = jest.spyOn(instance, "initializeModify");

            wrapper.setProps({ drawRegionMode: true });

            expect(spyRemoveInteraction).toBeCalledWith(instance.dragBox);
            expect(spyInitializeDraw).toBeCalledTimes(1);
            expect(spyAddInteraction).toBeCalledWith(instance.draw);
            expect(spyInitializeModify).toBeCalledTimes(1);
            expect(spyAddInteraction).toBeCalledWith(instance.modify);
            expect(spyAddInteraction).toBeCalledWith(instance.snap);
            expect(spyRemoveInteraction).toBeCalledWith(instance.draw);
            expect(spyRemoveInteraction).toBeCalledWith(instance.modify);
            expect(spyRemoveInteraction).toBeCalledWith(instance.snap);
        });

        it("should update interactions when pointer enters image.", () => {
            const spyAddInteraction = jest.spyOn(instance, "addInteraction");
            const spyRemoveInteraction = jest.spyOn(instance, "removeInteraction");
            const spyInitializeDraw = jest.spyOn(instance, "initializeDraw");
            const spyInitializeModify = jest.spyOn(instance, "initializeModify");

            wrapper.setProps({ drawRegionMode: true, isPointerOnImage: true, isSnapped: true, isDrawing: true });

            expect(spyRemoveInteraction).toBeCalledWith(instance.dragBox);
            expect(spyInitializeDraw).toBeCalledTimes(1);
            expect(spyAddInteraction).toBeCalledWith(instance.draw);
            expect(spyInitializeModify).toBeCalledTimes(1);
            expect(spyAddInteraction).toBeCalledWith(instance.modify);
            expect(spyAddInteraction).toBeCalledWith(instance.snap);
            expect(spyRemoveInteraction).toBeCalledWith(instance.draw);
            expect(spyRemoveInteraction).toBeCalledWith(instance.snap);
        });

        it("should update interactions when leaving draw region mode.", () => {
            const spyAddInteraction = jest.spyOn(instance, "addInteraction");
            const spyRemoveInteraction = jest.spyOn(instance, "removeInteraction");
            const spyInitializeModify = jest.spyOn(instance, "initializeModify");

            wrapper.setProps({ drawRegionMode: false, isPointerOnImage: false });

            expect(spyRemoveInteraction).toBeCalledWith(instance.draw);
            expect(spyAddInteraction).toBeCalledWith(instance.dragBox);
            expect(spyInitializeModify).toBeCalledTimes(1);
            expect(spyAddInteraction).toBeCalledWith(instance.modify);
            expect(spyAddInteraction).toBeCalledWith(instance.snap);
        });

        it("should update interactions when dragging vertex.", () => {
            const spyAddInteraction = jest.spyOn(instance, "addInteraction");
            const spyRemoveInteraction = jest.spyOn(instance, "removeInteraction");
            const spyInitializeModify = jest.spyOn(instance, "initializeModify");

            wrapper.setProps({ drawRegionMode: false, isPointerOnImage: false, isVertexDragging: true });

            expect(spyRemoveInteraction).toBeCalledWith(instance.draw);
            expect(spyAddInteraction).toBeCalledWith(instance.dragBox);
            expect(spyInitializeModify).toBeCalled();
            expect(spyAddInteraction).toBeCalledWith(instance.modify);
            expect(spyAddInteraction).toBeCalledWith(instance.snap);
            expect(spyRemoveInteraction).toBeCalledWith(instance.modify);
            expect(spyRemoveInteraction).toBeCalledWith(instance.dragBox);
        });
    });

    describe("Functions", () => {
        let wrapper: ReactWrapper;
        let instance;

        beforeEach(() => {
            const props = { ...baseProps, initEditorMap: true };
            wrapper = mount(<ImageMap {...props} />);
            instance = wrapper.instance() as any;
        });

        it("should reset layer visibility.", () => {
            instance.resetAllLayerVisibility();
            expect(instance.checkboxLayer.getVisible()).toBeTruthy();
            expect(instance.drawnLabelLayer.getVisible()).toBeTruthy();
            expect(instance.tableBorderLayer.getVisible()).toBeTruthy();
            expect(instance.tableIconLayer.getVisible()).toBeTruthy();
            expect(instance.tableIconBorderLayer.getVisible()).toBeTruthy();
            expect(instance.textLayer.getVisible()).toBeTruthy();
            expect(instance.drawnRegionLayer.getVisible()).toBeTruthy();
            expect(instance.podLayer.getVisible()).toBeTruthy();
        });

        it("should return resolution.", () => {
            expect(instance.getResolutionForZoom(0)).not.toBeNull();
        });

        it("should add feature to text layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addFeature(feature);
            expect(instance.getFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to text layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addFeatures(features);
            const returnedFeatures = instance.getAllFeatures();
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(features[0]);
        });

        it("should add feature to checkbox layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addCheckboxFeature(feature);
            expect(instance.getCheckboxFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to checkbox layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addCheckboxFeatures(features);
            const returnedFeatures = instance.getAllCheckboxFeatures();
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(features[0]);
        });

        it("should add feature to pod layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addPodFeature(feature);
            expect(instance.getPodFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to pod layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addPodFeatures(features);
            const returnedFeatures = instance.getAllPodFeatures();
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(features[0]);
        });

        it("should add feature to label layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addLabelFeature(feature);
            expect(instance.getLabelFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to label layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addLabelFeatures(features);
            const returnedFeatures = instance.getAllLabelFeatures();
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(features[0]);
        });

        it("should add feature to drawn label layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addDrawnLabelFeature(feature);
            expect(instance.getDrawnLabelFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to drawn label layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addDrawnLabelFeatures(features);
            const returnedFeatures = instance.getAllDrawnLabelFeatures();
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(features[0]);
        });

        it("should add feature to table border layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addTableBorderFeature(feature);
            expect(instance.getTableBorderFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to table border layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addTableBorderFeatures(features);

            // Note: imageMap doesn't expose getAllTableBorderFeatures.
            expect(instance.getTableBorderFeatureByID(region.id)).toBe(features[0]);
        });

        it("should add feature to table icon layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addTableIconFeature(feature);
            expect(instance.getTableIconFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to table icon layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addTableIconFeatures(features);

            // Note: imageMap doesn't expose getAllTableBorderFeatures.
            expect(instance.getTableIconFeatureByID(region.id)).toBe(features[0]);
        });

        it("should add feature to table icon border layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addTableIconBorderFeature(feature);
            expect(instance.getTableIconBorderFeatureByID(region.id)).toBe(feature);
        });

        it("should add features to table icon border layer.", () => {
            const region = makeMockRegion();
            const features = [convertRegionToFeature(region, mockExtent, false)];
            instance.addTableIconBorderFeatures(features);

            // Note: imageMap doesn't expose getAllTableIconBorderFeatures.
            expect(instance.getTableIconBorderFeatureByID(region.id)).toBe(features[0]);
        });

        it("should remove feature from text layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addFeature(feature);
            expect(instance.getFeatureByID(region.id)).toBe(feature);
            instance.removeFeature(feature);
            expect(instance.getFeatureByID(region.id)).toBeNull();
        });

        it("should remove feature from checkbox layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addCheckboxFeature(feature);
            expect(instance.getCheckboxFeatureByID(region.id)).toBe(feature);
            instance.removeCheckboxFeature(feature);
            expect(instance.getCheckboxFeatureByID(region.id)).toBeNull();
        });

        it("should remove feature from label layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addLabelFeature(feature);
            expect(instance.getLabelFeatureByID(region.id)).toBe(feature);
            instance.removeLabelFeature(feature);
            expect(instance.getLabelFeatureByID(region.id)).toBeNull();
        });

        it("should remove feature from drawn label layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addDrawnLabelFeature(feature);
            expect(instance.getDrawnLabelFeatureByID(region.id)).toBe(feature);
            instance.removeDrawnLabelFeature(feature);
            expect(instance.getDrawnLabelFeatureByID(region.id)).toBeNull();
        });

        it("should remove feature from drawn region layer.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);

            // Note: imageMap doesn't expose addDrawnRegionFeature.
            instance.addDrawnRegionFeatures([feature]);
            expect(instance.getDrawnRegionFeatureByID(region.id)).toBe(feature);
            instance.removeDrawnRegionFeature(feature);
            expect(instance.getDrawnRegionFeatureByID(region.id)).toBeNull();
        });

        it("should remove all features.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);

            instance.addFeature(feature);
            instance.addCheckboxFeature(feature);
            instance.addPodFeature(feature);
            instance.addLabelFeature(feature);
            instance.addTableBorderFeature(feature);
            instance.addTableIconFeature(feature);
            instance.addTableIconBorderFeature(feature);
            instance.addDrawnLabelFeature(feature);
            instance.addDrawnRegionFeatures([feature]);

            instance.removeAllFeatures();

            expect(instance.getFeatureByID(region.id)).toBeNull();
            expect(instance.getCheckboxFeatureByID(region.id)).toBeNull();
            expect(instance.getPodFeatureByID(region.id)).toBeNull();
            expect(instance.getLabelFeatureByID(region.id)).toBeNull();
            expect(instance.getTableBorderFeatureByID(region.id)).toBeNull();
            expect(instance.getTableIconFeatureByID(region.id)).toBeNull();
            expect(instance.getTableIconBorderFeatureByID(region.id)).toBeNull();
            expect(instance.getDrawnLabelFeatureByID(region.id)).toBeNull();
            expect(instance.getDrawnRegionFeatureByID(region.id)).toBeNull();
        });

        it("should remove all label features.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addLabelFeature(feature);
            instance.removeAllLabelFeatures();
            expect(instance.getLabelFeatureByID(region.id)).toBeNull();
        });

        it("should remove all pod features.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addPodFeature(feature);
            instance.removeAllPodFeatures();
            expect(instance.getPodFeatureByID(region.id)).toBeNull();
        });

        it("should remove all drawn label features.", () => {
            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addDrawnLabelFeature(feature);
            instance.removeAllDrawnLabelFeatures();
            expect(instance.getDrawnLabelFeatureByID(region.id)).toBeNull();
        });

        it("should remove/add interaction from the map.", () => {
            const interaction = new DragPan();
            const spyMapAddInteraction = jest.spyOn(instance.map, "addInteraction");
            const spyMapRemoveInteraction = jest.spyOn(instance.map, "removeInteraction");
            instance.removeInteraction(interaction);
            instance.addInteraction(interaction);
            expect(spyMapRemoveInteraction).toBeCalledTimes(1);
            expect(spyMapAddInteraction).toBeCalledTimes(1);
        });

        it("should update map size.", () => {
            const spyUpdateSize = jest.spyOn(instance, "updateSize");
            instance.updateSize();
            expect(spyUpdateSize).not.toThrow();
        });

        it("should return image extent.", () => {
            const extent = instance.getImageExtent();
            expect(extent.length).toEqual(4);
        });

        it("should return features in extent.", () => {
            const extent = instance.getImageExtent();
            expect(instance.getFeaturesInExtent(extent)).toEqual([]);

            const region = makeMockRegion();
            const feature = convertRegionToFeature(region, mockExtent, false);
            instance.addFeature(feature);
            const returnedFeatures = instance.getFeaturesInExtent(extent);
            expect(returnedFeatures.length).toEqual(1);
            expect(returnedFeatures[0]).toBe(feature);
        });

        it("should zoom in.", () => {
            const spyZoomIn = jest.spyOn(instance, "zoomIn");
            instance.zoomIn();
            expect(spyZoomIn).not.toThrow();
        });

        it("should zoom out.", () => {
            const spyZoomOut = jest.spyOn(instance, "zoomOut");
            instance.zoomOut();
            expect(spyZoomOut).not.toThrow();
        });

        it("should reset zoom.", () => {
            const spyResetZoom = jest.spyOn(instance, "resetZoom");
            instance.resetZoom();
            expect(spyResetZoom).not.toThrow();
        });

        it("should reset center.", () => {
            const spyResetCenter = jest.spyOn(instance, "resetCenter");
            instance.resetCenter();
            expect(spyResetCenter).not.toThrow();
        });

        it("should cancel modify.", () => {
            const spyCancelModify = jest.spyOn(instance, "cancelModify");
            instance.cancelModify();
            expect(spyCancelModify).not.toThrow();
        });

        it("should set swiping.", () => {
            instance.setSwiping(true);
            expect(instance.isSwiping).toBeTruthy();
            instance.setSwiping(false);
            expect(instance.isSwiping).not.toBeTruthy();
        });
    });

    describe("Events", () => {
        it("should handle mouse enters the map.", () => {
            const props = { ...baseProps, initEditorMap: true };
            const wrapper = mount(<ImageMap {...props} />);
            const mouseEnterHandler = jest.spyOn(wrapper.instance() as any, "handlePointerEnterImageMap");
            wrapper.simulate("mouseenter", { target: {} });
            expect(mouseEnterHandler).not.toThrow();
        });

        it("should handle mouse leaves the map.", () => {
            const props = {
                ...baseProps,
                initEditorMap: true,
                isDrawing: true,
                handleIsPointerOnImage: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const mouseLeaveHandler = jest.spyOn(wrapper.instance() as any, "handlePointerLeaveImageMap");
            wrapper.simulate("mouseleave", { target: {} });
            expect(mouseLeaveHandler).not.toThrow();
        });

        it("should handle pointer down the map.", () => {
            const props = {
                ...baseProps,
                initEditorMap: true,
                isPointerOnImage: true,
                enableFeatureSelection: true,
                handleIsPointerOnImage: noop,
                handleFeatureSelect: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(() => instance.handlePointerDown(makeBrowserEvent())).not.toThrow();
        });

        it("should handle pointer move the map.", () => {
            const props = {
                ...baseProps,
                initEditorMap: true,
                isPointerOnImage: true,
                enableFeatureSelection: true,
                handleIsPointerOnImage: noop,
                handleFeatureSelect: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(() => instance.handlePointerMove(makeBrowserEvent())).not.toThrow();
        });

        it("should handle pointer move the table icon.", () => {
            const props = {
                ...baseProps,
                initEditorMap: true,
                isPointerOnImage: true,
                enableFeatureSelection: true,
                handleIsPointerOnImage: noop,
                handleFeatureSelect: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(() => instance.handlePointerMoveOnFeatures(makeBrowserEvent())).not.toThrow();
        });

        it("should handle pointer up the map.", () => {
            const props = {
                ...baseProps,
                isDrawing: false,
                isVertexDragging: false,
                initEditorMap: true,
                isPointerOnImage: true,
                enableFeatureSelection: true,
                handleIsPointerOnImage: noop,
                handleFeatureSelect: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;

            instance.hasFeatureSelectedByPointer = false;
            expect(() => instance.handlePointerUp(makeBrowserEvent())).not.toThrow();
            expect(mockHandleFinishFeatureSelect).not.toBeCalled();

            instance.hasFeatureSelectedByPointer = true;
            expect(() => instance.handlePointerUp(makeBrowserEvent())).not.toThrow();
            expect(mockHandleFinishFeatureSelect).toBeCalledTimes(1);
        });

        it("should handle double-click the map.", () => {
            const props = {
                ...baseProps,
                initEditorMap: true,
                handleFeatureDoubleClick: noop,
            };
            const wrapper = mount(<ImageMap {...props} />);
            const instance = wrapper.instance() as any;
            expect(() => instance.handleDoubleClick(makeBrowserEvent())).not.toThrow();
        });
    });
});
