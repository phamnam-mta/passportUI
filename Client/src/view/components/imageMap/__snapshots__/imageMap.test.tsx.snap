// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageMap /> Rendering should match snapshot 1`] = `
<ImageMap
  checkboxFeatureStyler={[Function]}
  drawRegionStyler={[Function]}
  drawnRegionStyler={[Function]}
  featureStyler={[Function]}
  imageHeight={400}
  imageUri="test.jpg"
  imageWidth={300}
  labelFeatureStyler={[Function]}
  modifyStyler={[Function]}
  onFinishFeatureSelect={[MockFunction]}
  onMapReady={[Function]}
  tableBorderFeatureStyler={[Function]}
  tableIconBorderFeatureStyler={[Function]}
  tableIconFeatureStyler={[Function]}
>
  <div
    className="map-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      id="map"
      style={
        Object {
          "cursor": "default",
        }
      }
    />
  </div>
</ImageMap>
`;
